src/terminal/parser/InputStateMachineEngine.cpp

```diff
@@ -726,18 +797,127 @@ bool InputStateMachineEngine::_IsModified(const unsigned short cParams)
+// - rgusParams: the wchar_t to get the mapped vkey of. Represents the direction of the button (down vs up)
+// - pdwButtonState: Recieves the button state for the record
+// Return Value:
+// true iff we were able to synthesize pdwButtonState
+bool InputStateMachineEngine::_UpdateSGRMouseButtonState(const wchar_t wch,
+                                                         _In_reads_(cParams) const unsigned short* const rgusParams,
+                                                         _Out_ DWORD* const pdwButtonState,
+                                                         _Out_ DWORD* const pdwEventFlags)
+{
+    *pdwButtonState = _mouseButtonState;
+    *pdwEventFlags = 0;
+    bool fSuccess = true;
+
+    // We don't care about the middle 4 bits. So let's contatenate the top 2 bits and the bottom 2 bits from the 8 bit value
+    // The remaining bits represent which button had a change in state
+    const auto buttonID = (rgusParams[0] & 0x3) | (rgusParams[0] & 0xC0);
+
+    // Step 1: Translate which button was affected
+    DWORD buttonFlag = 0;
+    switch (buttonID)
+    {
+    case CsiMouseButtonCodes::Left:
+        buttonFlag = FROM_LEFT_1ST_BUTTON_PRESSED;
+        break;
+    case CsiMouseButtonCodes::Right:
+        buttonFlag = RIGHTMOST_BUTTON_PRESSED;
+        break;
+    case CsiMouseButtonCodes::Middle:
+        buttonFlag = FROM_LEFT_2ND_BUTTON_PRESSED;
+        break;
+    case CsiMouseButtonCodes::Button6:
+    case CsiMouseButtonCodes::Button7:
+    case CsiMouseButtonCodes::Button8:
+    case CsiMouseButtonCodes::Button9:
+    case CsiMouseButtonCodes::Button10:
+    case CsiMouseButtonCodes::Button11:
+    default:
+        fSuccess = false;
+        break;
+    }
+
+    // Step 2: Decide whether to set or clear that button's bit
+    // NOTE: WI_SetFlag/WI_ClearFlag can't be used herebecause buttonFlag would have to be a compile-time constant
+    switch (wch)
+    {
+    case CsiEndCodes::MOUSE_DOWN:
+        // set flag
+        *pdwButtonState |= buttonFlag;
+        break;
+    case CsiEndCodes::MOUSE_UP:
+        // clear flag
+        *pdwButtonState &= (~buttonFlag);
+        break;
+    default:
+        fSuccess = false;
+        break;
+    }
+
+    // Step 3: append the mouse wheel data to out param
+    switch (buttonID)
+    {
+    case CsiMouseButtonCodes::ScrollBack:
+    {
+        // set high word to proper scroll direction
+        // scroll intensity is assumed to be constant value
+        DWORD hi_word = static_cast<DWORD>(-128) << 4;
+        *pdwButtonState |= hi_word;
+
+        *pdwEventFlags |= MOUSE_WHEELED;
+        break;
+    }
+    case CsiMouseButtonCodes::ScrollForward:
+    {
+        // set high word to proper scroll direction
+        // scroll intensity is assumed to be constant value
+        DWORD hi_word = static_cast<DWORD>(128) << 4;
+        *pdwButtonState |= hi_word;
+
+        *pdwEventFlags |= MOUSE_WHEELED;
+        break;
+    }

```

Comment:
Woah wtf is happening here? Couldn't we pre-compute these bit masks and just have the literals in there? and use `SetFlag`, since the flags we're adding are constant?